_posts/2020-03-15-method-s-arguments-pt-1.markdown

```diff
@@ -0,0 +1,372 @@
+---
+layout: post
+title: Ruby Method's Arguments
+date: 2020-03-15 09:30:00
+categories: [ruby, learning]
+author: arieljuod
+---
+
+Ruby is an object oriented language where everything is an object (even methods are objects of the class Method!), so everything we need to do is done by calling methods on objects. That also means that methods have to provide a lot of flexibility because they are used everywhere.
+
+Ruby provides a lot of options to pass arguments to our methods, so we'll make this topic a series so it's not too long. We'll split the options in different categories and then break down everything with some examples and/or use cases.
+
+But first, we need to define the glosary:
+
+<!--more-->
+
+# Methods Vs. Functions
+
+Both methods and functions serve the same purpose: encapsulate a piece of code to reuse it with ease. And some times we use them as sinonyms, but there's a conceptual difference:
+
+- Functions return a result based only on the input, functions shouldn't depend on or modify the state of something that's not in the given inputs.
+- Methods are always executed in the context of an object, so we can call a method with no arguments but we always have access to the method's state and we can modify it.
+
+In Ruby, where everything is an object, we can never have real functions but we can have methods that won't use or modify the object's state if we want to have something similar.
+
+# Arguments Vs. Parameters
+
+Another thing that's worth to differenciate is the concept of arguments and parameters. When we define a method, the parameters are part of the definition, the parameters are the variables that will contain the values that we use when calling that method. Arguments are the actual values that we use when we call a method. So, in this code:
+
+```ruby
+def my_method(foo)
+  # foo is a parameter when defining the method
+end
+
+my_method("bar")
+# "baz" is the argument we use when calling the method
+```
+
+Most of the times I'll use the word "arguments" for simplicity, since a method defines the parameters and accepts arguments.
+
+# No arguments
+
+We can have methods that have no input, like this:
+
+```ruby
+def pi # parenthesis are optional and the keyword "return" is optional too
+  3.14
+end
+```
+
+But given that this is a method, it has access to the object's state to do different things
+
+```ruby
+class Circle
+  def initialize(radius)
+    @radius = radius
+  end
+
+  def pi
+    3.14
+  end
+
+  def area
+    pi * @radius ** 2
+  end
+end
+
+my_circle = Circle.new(3)
+puts my_circle.area
+# => 28.26
+
+other_circle = Circle.new(4)
+puts other_circle.area
+# => 50.24
+```
+
+# Positional arguments
+
+This type of arguments get that name because the order you use matters. You can have 0 or more positional arguments, and you can have required and optional arguments.
```

Comment:
```suggestion
This type of argument gets this name because the order you use matters. You can have 0 or more positional arguments, and you can have required and optional arguments.
```